### PR TITLE
Color TrackScheme Hierarchy according to the time point (Issue 213)

### LINKED ISSUE #213: Color TrackScheme Hierarchy according to the time point
# User Story

As a Mastodon user, I may use the TrackScheme Hierarchy view to get a generalised overview of the lineage tree. This view does not contain any individual spots and is not scaled according to the time model making it easier for the user to investigate the lineage tree structure as a whole.
It may still be useful for the user, if the lineage tree could be rendered in a way that highlights branches that already exist at a certain timepoint while navigating through the available time steps in other views.

# Acceptance criteria

1. The Branches in the TrackScheme Hierarchy that do not exist at a time point selected via BigDataViewer or TrackScheme are drawn in a faded color, e.g.: 
grafik
1. This feature is only available if the TrackScheme Hierarchy View is coupled to other views: 
grafik
1. Original branch color is used in this case: TimePointModel.TimePoint >= BranchSpot.FirstTimePoint
1. A mixture of background color and branch color is used in all other cases

# Tasks
- [x] Add a reference to the real time point model to the class MamutBranchViewTrackSchemeHierarchy in addition to the already existing reference to the dummy time point model
- [x] Use the reference to the real time point model (if exists) to color branches according to the current settings in the real time point model
- [x] Add option to toggle this behaviour on/off for TrackSchemeHierarchy View via Preferences

# Additional information:

* One TrackSchemePanel for views TrackScheme, TrackScheme Branch and TrackScheme Hierarchy
* The TrackSchemePanel has multiple Renderers: 
  * The Renderer for TrackScheme Hierarchy View is in the class PaintHierarchicalGraph
  * This class extends PaintGraph
  * PaintGraph is a Graph Renderer
* There is also the ModelGraph - this is the representation of the acutal lineage graph structure
* There is the TrackSchemeGraph. It is a Listener on the ModelGraph and defines it visualisation. The TrackSchemeGraph copies the ModelGraph. There is one TrackSchemeGraph per Window, thus enabling different TrackScheme colorings in different windows.
* There is the ModelBranchGraph. It is a Listener on the ModelGraph that contains the simplified structure of the ModelGraph that is used in the TrackSchemeHierarchy View
* There is a TimePointModel that contains the different timepoints of the Mastodon project
* The MamutBranchViewTrackSchemeHierarchy has a Dummy TimePointModel to enable its time independent visualisation
* The MamutBranchViewTrackSchemeHierarchy additionally needs some reference to the real time point model
* Each Spot is aware of its timepoint
* The MamutBranchViewTrackSchemeHierarchy extends MamutBranchViewTrackSchemeBranch (these are the windows)

### PR DESCRIPTION
**Pull request by stefanhahmann**
Closes #213

### CODE REVIEW

File: src/main/java/org/mastodon/views/trackscheme/LineageTreeLayoutImp.java
Comment (maarzt) [inline src/main/java/org/mastodon/views/trackscheme/LineageTreeLayoutImp.java:95]: Please remove the @Nonnulll annotation for the sake of consistency. All other fields are expected to be not null too.

File: src/main/java/org/mastodon/views/trackscheme/display/PaintGraph.java
Comment (maarzt) [inline src/main/java/org/mastodon/views/trackscheme/display/PaintGraph.java:530]: This method here is part of a hack. You inject the code that is actually needed for the PaintHierarchyGraph into this class but you dissable it by using the `isFadingGutureTimepoints()` method. Now there is a simple solution for this ScreenVertex.isFaded should always be false.

File: src/main/java/org/mastodon/views/trackscheme/display/style/TrackSchemeStyle.java
Comment (maarzt) [inline src/main/java/org/mastodon/views/trackscheme/display/style/TrackSchemeStyle.java:381]: Having this setting in the TrackSchemeStyle feels wrong to me. It is too much hidden from the user. And it also feels like a misuse of the TrackSchemeStyle. I think now with the FadedModel it is possible to have a icon/button in the MamutBranchViewTrackSchemeHierarchy window, that enables or disables the FadedModel accordingly.

File: src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java
Comment (maarzt) [inline src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java:155]: Removing this line will actually reintroduce a bug. Please don't remove that line. You may read the commit message that introduced this line to see what I'm talking about.

File: src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java
Comment (maarzt) [inline src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java:181]: Why do you call this timepointModel navigationTimepointModel? Please undo the renaming if there's no reasoning behind it.

File: src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java
Comment (maarzt) [inline src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java:206]: "optional" is not a good name. I agree. But the class is TrackSchemePanel. Please name the variable simply "options".

File: src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java:106]: I have removed it, since this is unneeded. Apologies for this mistake. Will take more care in the future.

File: src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java:206]: There is already a local variable in this method called "options". I followed your advise to rename the name of the variable in the method body to options, but I had to rename the existing variable "options" to "optionValues" to make this possible.

File: src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java:155]: Thanks for the hint. I added this line again together with a comment, explaining, why it is needed.

File: src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java:181]: There was a reasoning behind it during development. But that reason now does not exist anymore. Thus I undid the renaming.

File: src/main/java/org/mastodon/views/trackscheme/LineageTreeLayoutImp.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/views/trackscheme/LineageTreeLayoutImp.java:163]: Removed it.

File: src/main/java/org/mastodon/views/trackscheme/display/PaintGraph.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/views/trackscheme/display/PaintGraph.java:530]: I removed the method.

File: src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java:155]: I failed to reproduce the bug that is described here (https://github.com/mastodon-sc/mastodon/pull/201) when this line is removed. However, if change the line as you suggest, this introduces a bug into the fading.
Perhaps, you could show me how to reproduce the bug if that line is removed?

Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java:155]: After checking again, I could added this line again.
There was another error, which I solved by adding a second timepoint model. Now, there are two timepoint models, one that is for highlighting the current row of spots (which is overridden in trackschemehierarchy) and one that is used for computing the fading.

File: src/main/java/org/mastodon/views/trackscheme/display/TrackSchemeFrame.java
Comment (maarzt) [inline src/main/java/org/mastodon/views/trackscheme/display/TrackSchemeFrame.java:87]: The fadingTimepointModel isn't required here. Please add listeners to the fadedModel instead.

**Review by maarzt — CHANGES_REQUESTED** [review]

File: src/main/java/org/mastodon/app/ui/TimepointAndNumberOfSpotsLabel.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/app/ui/TimepointAndNumberOfSpotsLabel.java]: I agree. I removed it.

File: src/main/java/org/mastodon/app/ui/TimepointAndNumberOfSpotsLabel.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/app/ui/TimepointAndNumberOfSpotsLabel.java]: I changed it.
For further reference: why do you think a Panel having a Label is better than just a Label?

File: src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/views/trackscheme/display/TrackSchemePanel.java:463]: Thanks for this advice. I was also unsure about this line, since I did not fully understand all the different flags that currently exist.
I added fadingChanged and the corresponding flad.

File: src/main/java/org/mastodon/views/trackscheme/display/TrackSchemeFrame.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/views/trackscheme/display/TrackSchemeFrame.java:86]: Thanks for the advice. I had this idea as well but did not find a solution before.
Now, I tried again. It is solved now via listeners() and the fadingTimepointModel is not needed here anymore. However, I am not sure, if my implementation meets your intention. Please check.

Comment (stefanhahmann) [reply src/main/java/org/mastodon/views/trackscheme/display/TrackSchemeFrame.java]: After this discussing this again I introduced a FadingListener that is implemented by the TrackSchemePanel.

File: src/main/java/org/mastodon/app/ui/TimepointAndNumberOfSpotsLabel.java
Comment (maarzt) [reply src/main/java/org/mastodon/app/ui/TimepointAndNumberOfSpotsLabel.java]: It's the standard pattern to extend JPanel. Using JPanel is also a lot more flexible. Why would you want to extend JLabel?

File: src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/MamutBranchViewTrackScheme.java:155]: And now we finally don't even need the second timepoint model.

**Review by maarzt — APPROVED** [review]
I removed some unnecessary code, and moved a few other lines of code.
But overall the PR is ok now :relieved:

(I'm still worried about how many classes we needed to change to implement this one feature.)

**Comment by maarzt** [comment]
@tinevez I'm ok with this being merged now. But the changes are very complex. So please have a look first.

File: src/main/java/org/mastodon/app/ui/TimepointAndNumberOfSpotsLabel.java
Comment (stefanhahmann) [reply src/main/java/org/mastodon/app/ui/TimepointAndNumberOfSpotsLabel.java]: Keeping it simple.

My thought was to not use an extra JPanel where only a JLabel is actually needed. The potential extra flexibility of the JPanel is not needed the way this component is currently used. The JPanel in my view only adds an extra, yet unrequired, layer.

**Review by maarzt — CHANGES_REQUESTED** [review]
The PR somehow introduces a bug when rendering the the TrackScheme Branch.
The bug is triggered by resorting the TrackScheme using one of the plugins in mastodon-tomancak. For example "Plugins > Trees Management > Flip Descendants".

GIF 05 05 2023 14-45-05

This need to be clarified before merging.

**Comment by stefanhahmann** [comment]
> Ready for review?

Yes.
@maarzt reviewed already extensively, but since the changes are massive, a second review would be good.

**Review by maarzt — CHANGES_REQUESTED** [review]

File: src/test/java/org/mastodon/util/ColorUtilsTest.java
Comment (maarzt) [reply src/test/java/org/mastodon/util/ColorUtilsTest.java:58]: I removed the lines myself, and found another similar bug in `ColorUtils.mixColors(...)`.

Now there's a test to covers the alpha channel mixing in `ColorUtils.mixColors(...)`

**Review by maarzt — APPROVED** [review]

File: src/test/java/org/mastodon/util/ColorUtilsTest.java
Comment (stefanhahmann) [reply src/test/java/org/mastodon/util/ColorUtilsTest.java]: I was kind of anticipating that you judge this code as being misplaced here. For this reason, I had put it in a separate commit ([ColorUtils.mixColors](https://github.com/mastodon-sc/mastodon/pull/231/commits/119c095d02459ca6b86c2e2cfba76fe237a03e11)), so that this commit could just be dropped.

However: I would like to keep this code somewhere, as it contributed very much to solving the bug related to the scaleAlpha() method. It also demonstrates how the result of scaleAlpha should be used, but I confess the unit test, may not be the best place for this. Maybe an improved javadoc would also already help.

**Comment by stefanhahmann** [comment]
> Sorry I am late on this. Will work on it

@tinevez may we kindly remind you on this review? we would love to have this finished/merged.

**Comment by tinevez** [comment]
I made a PR for your PR with minor changes, mainly sizing the opportunity to fix the look of the settings panel. 
Take them if you feel like it.
To me otherwise it is good to be merged to dev.

**Review by tinevez — APPROVED** [review]

**Comment by stefanhahmann** [comment]
Hi @tinevez,

I merged your PR https://github.com/stefanhahmann/mastodon/pull/17.

However after merging the toolbar of the trackscheme windows looked like this on my (windows) machine:

grafik

Thus, I increased the size of the Group Locks Panel again: https://github.com/mastodon-sc/mastodon/pull/231/commits/316248399effe5b0d911e0022f2613f464874214

Also, I think, it is better to avoid abbreviations in user interfaces. Thus I made these changes: 

* https://github.com/mastodon-sc/mastodon/pull/231/commits/0deefa744d995f5ff45fdaaa515261cfa005e3f7
* https://github.com/mastodon-sc/mastodon/pull/231/commits/b54161374360039a49f51786da9a6e5ce4f600e5

which make the toolbar of trackscheme window look like this on my windows:

grafik

Hope you agree?

**Comment by tinevez** [comment]
Go for it!

**Comment by stefanhahmann** [comment]
> Go for it!

@maarzt may I ask you to merge (I do not have the rights to do so)? If you have questions regarding the remaining conflict, you may contact me.

**Comment by tinevez** [comment]
Merged! But I had to do it via a local branch I had rights on. 
This PR can be marked as merged and closed.

**Pull request closed without merging**
